### ISSUE TITLE
Add more Prometheus metrics

### ISSUE THREAD

**Issue by MichaelMure** [ipfs/go-ipfs#5604]
#### Version information:

0.4.17

#### Type:

Enhancement

#### Description:

Infura is hosting a set of IPFS nodes to provide a public IPFS API. We would like to monitor more metrics not currently exposed by go-ipfs as far as I can tell:

- bandwidth in/out (as given by `ipfs stats bw`), potentially split for the various subsystem (DHT, bitswap ...)
- repository size and object count (as given by `ipfs repo stat`)
- peers latency distribution (processed from `ipfs swarm peers --latency`)
- DHT metrics (latency, failure rate ...)

Is that something that would interest you ? I can work on that, but a few pointers would be appreciated.

CC @lgierth

**Comment by olizilla** [comment]
@Kubuxu can you explaina a little more about how we might set up those additional metrics?

I'd like to add an `ipfs_info` metric as a gauge with labels for `version` and `commit` as per https://www.robustperception.io/exposing-the-software-version-to-prometheus

It seems to be a common pattern, we already expose go version info like

```
# HELP go_info Information about the Go environment.
# TYPE go_info gauge
go_info{version="go1.12.7"} 1
```

Do we need to update https://github.com/ipfs/go-metrics-interface to support labels? Or can you show how I would do it?

Also, what's going on here in corehttp/metrics.go https://github.com/ipfs/go-ipfs/blob/9ea7c6a11113bebd0efed9b7d5a23331ed4f93b0/core/corehttp/metrics.go#L96-L106

They seem to be using the prometheus lib directly, but are the only examples I can find of metrics that use labels.

**Comment by Kubuxu** [comment]
For that, I would use Prometheus directly (or OpenCensus/OpenMetric or whatever it is called now).
I didn't follow recent metric collection discussions, so @Stebalien might have other recommendations.

**Comment by requilence** [comment]
@MichaelMure just want to reach out. In case you have any progress on this enhancement we can continue to work on it to merge into go-ipfs

**Comment by MichaelMure** [comment]
Sorry I don't, I moved to other topics within the IPFS realm.

**Comment by varshadqz** [comment]
Can we get more metrics like peers and files information from http://localhost:5001/debug/metrics/prometheus?

**Comment by varshadqz** [comment]
Please guide how could I add more ipfs metrics?

**Comment by Stebalien** [comment]
@varshadqz could you be a bit more specific?

**Comment by varshadqz** [comment]
I need the ipfs metrics like peers and file information as shown on the ipfs desktop which I want to display using grafana. How can I get them?

**Comment by Stebalien** [comment]
All currently available metrics can be seen at http://localhost:5001/debug/metrics/prometheus. Additionally, you may want to poll the API itself to collect additional metrics.

If you want go-ipfs to expose additional metrics, please be _very_ specific about the metrics you need.

**Comment by varshadqz** [comment]
@Stebalien I need metrics as below:
1. ipfs status (gateway, peer id, public key)
2. ipfs peers ids
3. total files and repo
4. the total size of all files and repo
5.  list and count of Pinned files 
6. blocks count

**Comment by Stebalien** [comment]
For that kind of info, you'd be better served scraping the API yourself (possibly writing a custom prometheus exporter). Listing pins, peers, etc over the metrics API will make the metrics API unusable (and really slow).

See https://www.robustperception.io/writing-json-exporters-in-python for an example.

**Comment by varshadqz** [comment]
Thank you for the reply and suggestion @Stebalien.

**Comment by varshadqz** [comment]
Hi @Stebalien may I know about this existed metrics` ipfs_fsrepo_datastore_getsize_total`. Is this gives the total repositories size of connected ipfs?

**Comment by Stebalien** [comment]
No, that's the number of types the GetSize method has been called. You want the `RepoSize` field of `curl -X POST 'http://localhost:5001/api/v0/repo/stat?size-only=true'`.

**Comment by varshadqz** [comment]
Oh, thank you. If I run the command `curl -X POST 'http://localhost:5001/api/v0/repo/stat?size-only=false'`, it returns `{"RepoSize":8680297,"StorageMax":10000000000,"NumObjects":1692,"RepoPath":"/home/varshakhandre14/.ipfs","Version":"fs-repo@11"}` So, `NumObjects` are the total number of directories of my local ipfs with the connected peers have? Or directories of only my local ipfs?

**Comment by Stebalien** [comment]
So, `NumObjects` are the total number of directories of my local ipfs with the connected peers have? Or directories of only my local ipfs?

Neither. It's the number of "data chunks" your node has. Files are chunked in pieces.

**Comment by varshadqz** [comment]
Okay. Thank you!

**Comment by varshadqz** [comment]
Hi @Stebalien, If I need the total size of files that exists in ipfs, do I need `CumulativeSize` in  ` curl -X POST "http://127.0.0.1:5001/api/v0/files/stat?arg="/""`  ?

**Comment by Stebalien** [comment]
Could you open a thread on discuss.ipfs.io to discuss this further (feel free to tag me there as @Stebalien). We're getting a bit off topic for this issue.
